src/charm.py

```diff
@@ -94,6 +115,35 @@ def _pebble_layer(self) -> Dict:
             },
         }
 
+    def _reset_instance_action(self, event: ActionEvent) -> None:
+        """Reset instance and report action result.
+
+        Args:
+            event: Event triggering the reset instance action.
+        """
+        results = {
+            "reset-instance": False,
+        }
+        if not self.model.unit.is_leader():
+            event.defer()
+            return
+        container = self.unit.get_container(SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            event.fail("Failed to connect to container")
+            return
+        self.model.unit.status = ops.MaintenanceStatus("Resetting Synapse instance")
+        try:
+            self._synapse.reset_instance_action(container)
+            self._synapse.execute_migrate_config(container)
```

Comment:
Should this be a single function on `synapse`? Perhaps a function that just calls these two methods?

Reply:
I see your point but since execute_migrate_config is executed here and also on config-changed, I believe would be better for now to let them explicitly separated.